rtCommon/bidsArchive.py

```diff
@@ -0,0 +1,825 @@
+    def _imagesAppendCompatible(img1: nib.Nifti1Image,
+                                img2: nib.Nifti1Image) -> Tuple[bool, str]:
+        """
+        Verifies that two Nifti image headers match in along a defined set of
+        NIfTI header fields which should not change during a continuous fMRI
+        scanning session.
+
+        This is primarily intended as a safety check, and does not conclusively
+        determine that two images are valid to append to together or are part of
+        the same scanning session.
+
+        Args:
+            header1: First Nifti header to compare (dict of numpy arrays)
+            header2: Second Nifti header to compare (dict of numpy arrays)
+
+        Returns:
+            True if the headers match along the required dimensions, false
+            otherwise.
+
+        """
+        fieldsToMatch = ["intent_p1", "intent_p2", "intent_p3", "intent_code",
+                         "dim_info", "datatype", "bitpix",
+                         "slice_duration", "toffset", "scl_slope", "scl_inter",
+                         "qform_code", "quatern_b", "quatern_c", "quatern_d",
+                         "qoffset_x", "qoffset_y", "qoffset_z",
+                         "sform_code", "srow_x", "srow_y", "srow_z"]
+
+        header1 = img1.header
+        header2 = img2.header
+
+        for field in fieldsToMatch:
+            v1 = header1.get(field)
+            v2 = header2.get(field)
+
+            # Use slightly more complicated check to properly match nan values
+            if not (np.allclose(v1, v2, atol=0.0, equal_nan=True)):
+                errorMsg = (f"NIfTI headers don't match on field: {field} "
+                            f"(v1: {v1}, v2: {v2})")
+                return (False, errorMsg)
+
+        # Two NIfTI headers are append-compatible in 2 cases:
+        # 1) Pixel dimensions are exactly equal, and dimensions are equal except
+        # for in the final dimension
+        # 2) One image has one fewer dimension than the other, and all shared
+        # dimensions and pixel dimensions are exactly equal
+        dimensionMatch = True
+
+        dimensions1 = header1.get("dim")
+        dimensions2 = header2.get("dim")
+
+        nDimensions1 = dimensions1[0]
+        nDimensions2 = dimensions2[0]
+
+        pixdim1 = header1.get("pixdim")
+        pixdim2 = header2.get("pixdim")
+
+        # Case 1
+        if nDimensions1 == nDimensions2:
+            pixdimEqual = np.array_equal(pixdim1, pixdim2)
+            allButFinalEqual = np.array_equal(dimensions1[:nDimensions1],
+                                              dimensions2[:nDimensions2])
+
+            if not (pixdimEqual and allButFinalEqual):
+                dimensionMatch = False
+        # Case 2
+        else:
+            dimensionsDifferBy1 = abs(nDimensions1 - nDimensions2) == 1
+
+            nSharedDimensions = min(nDimensions1, nDimensions2)
+            # Arrays are 1-indexed as # dimensions is stored in first slot
```

Comment:
GW comment: maybe something like this and the next comment duplicated up higher as well

Reply:
TODO: Transfer the code to a comment